x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanStreamInput.java

```diff
@@ -28,16 +28,34 @@
 
 import java.io.IOException;
 import java.util.Collection;
+import java.util.HashMap;
 import java.util.HashSet;
+import java.util.Map;
 import java.util.function.LongFunction;
+import java.util.function.Supplier;
 
 /**
  * A customized stream input used to deserialize ESQL physical plan fragments. Complements stream
  * input with methods that read plan nodes, Attributes, Expressions, etc.
  */
 public final class PlanStreamInput extends NamedWriteableAwareStreamInput {
 
-    private static final LongFunction<NameId> DEFAULT_NAME_ID_FUNC = NameId::new;
+    /**
+     * A Mapper of stream named id, represented as a primitive long value, to NameId instance.
+     * The no-args NameId constructor is used for absent entries, as it will automatically select
+     * and increment an id from the global counter, thus avoiding potential conflicts between the
+     * id in the stream and id's during local re-planning on the data node.
+     */
+    static final class NameIdMapper implements LongFunction<NameId> {
+        final Map<Long, NameId> seen = new HashMap<>();
+
+        @Override
+        public NameId apply(long streamNameId) {
+            return seen.computeIfAbsent(streamNameId, k -> new NameId());
```

Comment:
Since we no longer use `NameId(long)`constructor, should we remove it from the NameId?